### PR TITLE
[Refactor] fix peek semantic in `SharedBufferedInputStream` and `CacheInputStream`

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

`peek` interface is not supposed to advance read offset.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by dirtysalt** [comment]
run starrocks_admit_test

File: be/src/formats/parquet/column_chunk_reader.cpp [hunk 1]
```diff
@@ -140,6 +140,9 @@ Status ColumnChunkReader::_read_and_decompress_page_data(uint32_t compressed_siz
     Slice read_data;
     auto ret = _page_reader->peek(read_size);
     if (ret.ok()) {
+        // peek dos not advance offset.
+        int64_t pos = _page_reader->get_offset();
+        _page_reader->seek_to_offset(pos + read_size);
```
Comment (sduzh) [inline be/src/formats/parquet/column_chunk_reader.cpp:145]:
```suggestion
    RETURN_IF_ERROR(_page_reader->skip(ret->size());
```
The number of bytes actually returned by `peek()` may be less than the requested size, i.e, you should not assume `ret->size() == read_size` is always true.

File: be/src/formats/parquet/column_chunk_reader.cpp [hunk 2]
```diff
@@ -140,6 +140,9 @@ Status ColumnChunkReader::_read_and_decompress_page_data(uint32_t compressed_siz
     Slice read_data;
     auto ret = _page_reader->peek(read_size);
     if (ret.ok()) {
+        DCHECK_EQ(ret.value().size, read_size);
```
Comment (sduzh) [inline be/src/formats/parquet/column_chunk_reader.cpp:143]: Are you sure this will always be true under your implementations?

Comment (dirtysalt) [reply be/src/formats/parquet/column_chunk_reader.cpp:143]: yes. if peekable, then always true.

Comment (sduzh) [reply be/src/formats/parquet/column_chunk_reader.cpp:143]: Ok, please note this is implementation-defined, there is no guarantee that all future `peek()` implementations will behave in the same way.

Comment (sduzh) [reply be/src/formats/parquet/column_chunk_reader.cpp:143]: But this is not defined by the interface, I think we should coding to the interface but not the implementation. IMO, we should either change the usage or the interface definition.

Comment (sduzh) [reply be/src/formats/parquet/column_chunk_reader.cpp:143]: > But this is not defined by the interface, I think we should coding to the interface but not the implementation. IMO, we should either change the usage or the interface definition.

I prefer to update the usage.

Comment (dirtysalt) [reply be/src/formats/parquet/column_chunk_reader.cpp:143]: OK, I'll update it.

**Review by sduzh — DISMISSED** [review]

**Comment by dirtysalt** [comment]
run sr-admin-test

**Comment by dirtysalt** [comment]
run starrocks-admin-test

**Review by Youngwb — DISMISSED** [review]

**Review by murphyatwork — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**